dangerzone/isolation_provider.py

```diff
@@ -0,0 +1,501 @@
+                success = True
+
+        # Clean up
+        tmpdir.cleanup()
+
+        return success
+
+    def get_max_parallel_conversions(self) -> int:
+
+        # FIXME hardcoded 1 until timeouts are more limited and better handled
+        # https://github.com/freedomofpress/dangerzone/issues/257
+        return 1
+
+        n_cpu = 1  # type: ignore [unreachable]
+        if platform.system() == "Linux":
+            # if on linux containers run natively
+            cpu_count = os.cpu_count()
+            if cpu_count is not None:
+                n_cpu = cpu_count
+
+        elif self._get_runtime_name() == "docker":
+            # For Windows and MacOS containers run in VM
+            # So we obtain the CPU count for the VM
+            n_cpu_str = subprocess.check_output(
+                [self.get_runtime(), "info", "--format", "{{.NCPU}}"],
+                text=True,
+                startupinfo=get_subprocess_startupinfo(),
+            )
+            n_cpu = int(n_cpu_str.strip())
+
+        return 2 * n_cpu + 1
+
+
+class Dummy(IsolationProvider):
+    """Dummy Isolation Provider (FOR TESTING ONLY)
+
+    "Do-nothing" converter - the sanitized files are the same as the input files.
+    Useful for testing without the need to use docker.
+    """
+
+    def __init__(self) -> None:
+        # Sanity check
+        if not getattr(sys, "dangerzone_dev", False):
+            raise Exception(
+                "Dummy isolation provider is UNSAFE and should never be "
+                + "called in a non-testing system."
+            )
+
+    def install(self) -> bool:
+        pass
+
+    def _convert(
+        self,
+        document: Document,
+        ocr_lang: Optional[str],
+        stdout_callback: Optional[Callable] = None,
+    ) -> bool:
+        log.debug("Dummy converter started:")
+        log.debug(
+            f"  - document: {os.path.basename(document.input_filename)} ({document.id})"
+        )
+        log.debug(f"  - ocr     : {ocr_lang}")
+        log.debug("\n(simulating conversion)")
+
+        success = True
+
+        progress = [
+            [False, "Converting to PDF using GraphicsMagick", 0.0],
+            [False, "Separating document into pages", 3.0],
+            [False, "Converting page 1/1 to pixels", 5.0],
+            [False, "Converted document to pixels", 50.0],
+            [False, "Converting page 1/1 from pixels to PDF", 50.0],
+            [False, "Merging 1 pages into a single PDF", 95.0],
+            [False, "Compressing PDF", 97.0],
+            [False, "Safe PDF created", 100.0],
+        ]
+
+        for (error, text, percentage) in progress:
+            self._print_progress(document, error, text, percentage)  # type: ignore [arg-type]
+            if stdout_callback:
+                stdout_callback(error, text, percentage)
+            if error:
+                success = False
+            time.sleep(0.2)
+
+        return success
+
+    def _print_progress(
```

Comment:
We can move this method to `IsolationProvider`, and use it both from the `Container` and `Dummy` classes. Even though the class is an abstract one, it's legal to have working methods in it.

Reply:
Good point. Done in [df6231d](https://github.com/freedomofpress/dangerzone/pull/302/commits/df6231d4464c02eb28a1cff22ff4e63a1cf0aa08).